README.md
# HTML::Pipeline::LinkifyGitHub [![Build Status](https://travis-ci.org/jollygoodcode/html-pipeline-linkify_github.svg)](https://travis-ci.org/jollygoodcode/html-pipeline-linkify_github)

`HTML::Pipeline::LinkifyGitHub` provides a [HTML::Pipeline](https://github.com/jch/html-pipeline)
filter to autolink GitHub urls.

## Installation

Add this line to your application's Gemfile:

```ruby
gem "html-pipeline-linkify_github"
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install html-pipeline-linkify_github

## Usage

**Use `HTML::Pipeline::LinkifyGitHubFilter` filter after your markdown filter.**

```ruby
require "html/pipeline"
require "html/pipeline/linkify_github"

pipeline = HTML::Pipeline.new [
  HTML::Pipeline::MarkdownFilter,
  HTML::Pipeline::LinkifyGitHubFilter
]

result = pipeline.call <<-MARKDOWN.strip_heredoc
  https://github.com/rails/rails/pull/21862
  https://github.com/rails/rails/issues/21843
  https://github.com/rails/rails/commit/67597e1719ec6af7e22964603cc77aa5b085a864
MARKDOWN

puts result[:output].to_html
```

prints:

```html
<p><a href="https://github.com/rails/rails/pull/21862">rails/rails#21862</a><br>
<a href="https://github.com/rails/rails/issues/21843">rails/rails#21843</a><br>
<a href="https://github.com/rails/rails/commit/67597e1719ec6af7e22964603cc77aa5b085a864">rails/rails@`67597e`</a></p>
```

## Notes

This RubyGem requires Ruby 2.0+ because [support for 1.9.3 is officially end](https://www.ruby-lang.org/en/news/2014/01/10/ruby-1-9-3-will-end-on-2015/).

## Contributing

Please see the [CONTRIBUTING.md](/CONTRIBUTING.md) file.

## Credits

A huge THANK YOU to all our [contributors](https://github.com/jollygoodcode/html-pipeline-linkify_github/graphs/contributors)! :heart:

## License

Please see the [LICENSE.md](/LICENSE.md) file.

## Maintained by Jolly Good Code

[![Jolly Good Code](https://cloud.githubusercontent.com/assets/1000669/9362336/72f9c406-46d2-11e5-94de-5060e83fcf83.jpg)](http://www.jollygoodcode.com)

We specialise in Agile practices and Ruby, and we love contributing to open source.
[Speak to us](http://www.jollygoodcode.com/#get-in-touch) about your next big idea, or [check out our projects](http://www.jollygoodcode.com/open-source).


CHANGELOG.md
# CHANGELOG

## Unreleased

## 1.0.2 - 2015.10.20

- Fix children tags got removed when converting urls. [#4](https://github.com/jollygoodcode/html-pipeline-linkify_github/pull/4)

## 1.0.1 - 2015.10.15

- Linkify after markdown generated html. [#3](https://github.com/jollygoodcode/html-pipeline-linkify_github/pull/3/)

## 1.0.0 - 2015.10.04

Released v1.0.0.


test/test_helper.rb
require "bundler/setup"
require "minitest/autorun"
require "html/pipeline/linkify_github"

module LinkifyGitHubTestExtensions
  def assert_equal_html(expected, actual)
    assert_equal Nokogiri::HTML::DocumentFragment.parse(expected).to_hash,
                 Nokogiri::HTML::DocumentFragment.parse(actual).to_hash
  end
end

class Minitest::Test
  include LinkifyGitHubTestExtensions
end

class String
  def strip_heredoc
    gsub(/^#{scan(/^[ \t]*(?=\S)/).min}/, "".freeze)
  end
end


test/linkify_github_test.rb
require "test_helper"

class HTML::Pipeline::LinkifyGitHubTest < Minitest::Test
  LinkifyGitHubFilter = HTML::Pipeline::LinkifyGitHubFilter

  def filter(html)
    LinkifyGitHubFilter.call(html)
  end

  def test_that_it_has_a_version_number
    assert HTML::Pipeline::LinkifyGitHub::VERSION
  end

  def pull_request_url
    "https://github.com/rails/rails/pull/21862"
  end

  def issue_url
    "https://github.com/rails/rails/issues/21843"
  end

  def commit_url
    "https://github.com/rails/rails/commit/67597e1719ec6af7e22964603cc77aa5b085a864"
  end

  def test_linkify_github_pull_request_html_link
    body = %Q(<a href="#{pull_request_url}">#{pull_request_url}</a>)
    doc  = Nokogiri::HTML::DocumentFragment.parse(body)

    res = filter(doc)
    assert_equal_html %Q(<a href="#{pull_request_url}">rails/rails#21862</a>),
                      res.to_html
  end

  def test_linkify_github_issue_html_link
    body = %Q(<a href="#{issue_url}">#{issue_url}</a>)
    doc  = Nokogiri::HTML::DocumentFragment.parse(body)

    res = filter(doc)
    assert_equal_html %Q(<a href="#{issue_url}">rails/rails#21843</a>),
                      res.to_html
  end

  def test_linkify_github_commit_html_link
    body = %Q(<a href="#{commit_url}">#{commit_url}</a>)
    doc  = Nokogiri::HTML::DocumentFragment.parse(body)

    res = filter(doc)
    assert_equal_html %Q(<a href="#{commit_url}">rails/rails@<code>67597e</code></a>),
                      res.to_html
  end
end


test/linkify_github_integration_test.rb
require "test_helper"
require "html/pipeline"
require "html/pipeline/linkify_github"
require "github/markdown"

class HTML::Pipeline::LinkifyGitHubIntegrationTest < Minitest::Test
  def pipeline
    @pipeline ||= HTML::Pipeline.new [
      HTML::Pipeline::MarkdownFilter,
      HTML::Pipeline::LinkifyGitHubFilter
    ]
  end

  def test_works_with_markdown_filter
    result = pipeline.call <<-MARKDOWN.strip_heredoc
      https://github.com/rails/rails/pull/21862
      https://github.com/rails/rails/issues/21843
      https://github.com/rails/rails/commit/67597e1719ec6af7e22964603cc77aa5b085a864
    MARKDOWN

    assert_equal "<p><a href=\"https://github.com/rails/rails/pull/21862\">rails/rails#21862</a><br>\n<a href=\"https://github.com/rails/rails/issues/21843\">rails/rails#21843</a><br>\n<a href=\"https://github.com/rails/rails/commit/67597e1719ec6af7e22964603cc77aa5b085a864\">rails/rails@<code>67597e</code></a></p>",
                 result[:output].to_html
  end

  def test_works_when_markdown_already_linkified
    result = pipeline.call <<-MARKDOWN.strip_heredoc
      - [rails/rails#21862](https://github.com/rails/rails/pull/21862)
    MARKDOWN

    assert_equal "<ul>\n<li><a href=\"https://github.com/rails/rails/pull/21862\">rails/rails#21862</a></li>\n</ul>",
                 result[:output].to_html
  end

  def test_preserve_tags_inside_link
    result = pipeline.call <<-MARKDOWN.strip_heredoc
      - [**rails/rails#21862**](https://github.com/rails/rails/pull/21862)
    MARKDOWN

    assert_equal "<ul>\n<li><a href=\"https://github.com/rails/rails/pull/21862\"><strong>rails/rails#21862</strong></a></li>\n</ul>",
                 result[:output].to_html
  end
end


lib/html/pipeline/linkify_github.rb
require "html/pipeline"

module HTML
  class Pipeline
    # Filter that converts GitHub's url into friendly markdown.
    #
    # For example:
    #
    #   <a href="https://github.com/rails/rails/pull/21862">https://github.com/rails/rails/pull/21862</a>
    #   <a href="https://github.com/rails/rails/issues/21843">https://github.com/rails/rails/issues/21843</a>
    #   <a href="https://github.com/rails/rails/commit/67597e1719ec6af7e22964603cc77aa5b085a864">https://github.com/rails/rails/commit/67597e1719ec6af7e22964603cc77aa5b085a864</a>
    #
    #   =>
    #
    #   <a href="https://github.com/rails/rails/pull/21862">rails/rails#21862</a>
    #   <a href="https://github.com/rails/rails/issues/21843">rails/rails#21843</a>
    #   <a href="https://github.com/rails/rails/commit/67597e1719ec6af7e22964603cc77aa5b085a864">rails/rails@`67597e`</a>
    #
    # This filter does not write any additional information to the context hash.
    class LinkifyGitHubFilter < Filter
      PULL_REQUEST_REGEXP = %r{https?://(www.)?github.com/(?<owner>.+)/(?<repo>.+)/pull/(?<number>\d+)/?}.freeze
      ISSUES_REGEXP = %r{https?://(www.)?github.com/(?<owner>.+)/(?<repo>.+)/issues/(?<number>\d+)/?}.freeze
      COMMIT_REGEXP = %r{https?://(www.)?github.com/(?<owner>.+)/(?<repo>.+)/commit/(?<number>\w+)/?}.freeze

      def call
        doc.search("a").each do |element|
          next if element.blank? || element.comment?
          next if element["href"].to_s.empty?

          text = element.inner_html

          element.inner_html = if is_a_pull_request_link? text
            replace_pull_request_link(text)
          elsif is_a_issue_link? text
            replace_issue_link(text)
          elsif is_a_commit_link? text
            replace_commit_link(text)
          else
            text
          end
        end

        doc
      end

      private

        def is_a_pull_request_link?(text)
          text.include?("/pull/".freeze)
        end

        def is_a_issue_link?(text)
          text.include?("/issues/".freeze)
        end

        def is_a_commit_link?(text)
          text.include?("/commit/".freeze)
        end

        def replace_pull_request_link(text)
          text.gsub(PULL_REQUEST_REGEXP) do
            pull_request_shorthand($1, $2, $3)
          end
        end

        def replace_issue_link(text)
          text.gsub(ISSUES_REGEXP) do
            issue_shorthand($1, $2, $3)
          end
        end

        def replace_commit_link(text)
          text.gsub(COMMIT_REGEXP) do
            commit_shorthand($1, $2, $3)
          end
        end

        def pull_request_shorthand(repo, owner, number)
          "#{repo}/#{owner}##{number}"
        end

        def issue_shorthand(repo, owner, number)
          "#{repo}/#{owner}##{number}"
        end

        def commit_shorthand(repo, owner, number)
          "#{repo}/#{owner}@<code>#{number[0..5]}</code>"
        end
    end
  end
end
